db/range_del_aggregator_v2.cc

```diff
@@ -0,0 +1,189 @@
+        parsed_largest.sequence == kMaxSequenceNumber) {
+      // The file boundary has been artificially extended by a range tombstone.
+      // We do not need to adjust largest to properly truncate range
+      // tombstones that extend past the boundary.
+    } else if (parsed_largest.sequence == 0) {
+      // No range tombstone from this sstable can cover largest (or a range
+      // tombstone has seqnum 0 and does not cover any keys). We will never use
+      // largest to truncate a tombstone, so leave it as-is.
+    } else {
+      // The same user key may straddle two sstable boundaries. To ensure that
+      // the truncated end key can cover the largest key in this sstable, reduce
+      // its sequence number by 1.
+      parsed_largest.sequence -= 1;
+    }
+    largest_ = &parsed_largest;
+  }
+}
+
+bool TruncatedRangeDelIterator::Valid() const {
+  return iter_->Valid() &&
+         (smallest_ == nullptr ||
+          icmp_->Compare(*smallest_, iter_->parsed_end_key()) < 0) &&
+         (largest_ == nullptr ||
+          icmp_->Compare(iter_->parsed_start_key(), *largest_) < 0);
+}
+
+void TruncatedRangeDelIterator::Next() { iter_->TopNext(); }
+
+void TruncatedRangeDelIterator::Prev() { iter_->TopPrev(); }
+
+// NOTE: target is a user key
+void TruncatedRangeDelIterator::Seek(const Slice& target) {
+  if (largest_ != nullptr &&
+      icmp_->Compare(*largest_, ParsedInternalKey(target, kMaxSequenceNumber,
+                                                  kTypeRangeDeletion)) <= 0) {
+    iter_->Invalidate();
+    return;
+  }
+  iter_->Seek(target);
+}
+
+// NOTE: target is a user key
+void TruncatedRangeDelIterator::SeekForPrev(const Slice& target) {
+  if (smallest_ != nullptr &&
+      icmp_->Compare(ParsedInternalKey(target, 0, kTypeRangeDeletion),
+                     *smallest_) < 0) {
+    iter_->Invalidate();
+    return;
+  }
+  iter_->SeekForPrev(target);
+}
+
+void TruncatedRangeDelIterator::SeekToFirst() { iter_->SeekToTopFirst(); }
+
+void TruncatedRangeDelIterator::SeekToLast() { iter_->SeekToTopLast(); }
+
+RangeDelAggregatorV2::RangeDelAggregatorV2(const InternalKeyComparator* icmp,
+                                           SequenceNumber upper_bound)
+    : icmp_(icmp), upper_bound_(upper_bound) {}
+
+void RangeDelAggregatorV2::AddTombstones(
+    std::unique_ptr<FragmentedRangeTombstoneIterator> input_iter,
+    const InternalKey* smallest, const InternalKey* largest) {
+  if (input_iter == nullptr || input_iter->empty()) {
+    return;
+  }
+  if (wrapped_range_del_agg != nullptr) {
+    wrapped_range_del_agg->AddTombstones(std::move(input_iter), smallest,
+                                         largest);
+    // TODO: this eats the status of the wrapped call; may want to propagate it
+    return;
+  }
+  iters_.emplace_back(new TruncatedRangeDelIterator(std::move(input_iter),
+                                                    icmp_, smallest, largest));
```

Comment:
Yes, we had some discussion around this. Right now there is some difficulty because a file's range tombstones are added to the aggregator simply when a regular (point-key) `LevelIterator` opens that file. However we cannot analogously drop tombstones from the aggregator when the `Leveliterator` advances to the next file because of cases like the following:

- L1 has file1 containing key `"a"` and range tombstone `["b", "c")`
- L1 has file2 containing key `"d"`
- L2 has file3 containing key `"b"`
- User calls `Seek("b")`.
  - The L1 `LevelIterator` opens file1 and adds its range tombstones to the aggregator.
  - The L1 `LevelIterator` doesn't find any key greater than or equal to the lookup key, so it proceeds to close file1 and open file2. When file2 is opened, its range tombstones are also added to the aggregator.
  - The L2 `LevelIterator` opens file3 and finds key `"b"`. It calls `ShouldDelete` and the range tombstone from file1 must still be in the aggregator.

I believe @abhimadan has a proposal involving active and inactive lists, and tracking file boundaries, to overcome this obstacle without having to restructure how we add tombstones to the aggregator. Will let him write it :).

Reply:
Thanks for the intro @ajkr :)

In this PR, we don't keep track of iterator positions, so we can't make any optimizations. However, in #4677, we keep track of the following data structures (in the forward case; the reverse case is similar so I won't describe it here):
- an active iterator min-heap ordered by end key, which contains iterators which currently point to tombstones covering the ShouldDelete key
- an inactive iterator min-heap ordered by start key, which contains iterators which currently point to tombstones starting after the ShouldDelete key
- a consumed list, which contains iterators whose tombstones all end before the ShouldDelete key

This way, we only need to seek in the inactive and active iterators, while still allowing data iterators to add range tombstones to the aggregator as necessary. This improves performance for longer range scans, but in the current implementation, we don't free consumed iterators. I think that will require some more thought, though I have a few ideas on how to approach it (e.g., using file boundaries to determine whether an iterator can be cleaned up or not).

Reply:
I think there are a couple things we're conflating:

(1) having `BlockBasedTableIterator` return a sentinel key for the beginning/end of a file when it's extended by range tombstones so it'll be closed at the same time the tombstones become irrelevant; and
(2) having all range tombstones inlined in the data blocks.

It looks like we're mostly talking about the downsides of (2), but what about (1)?

Reply:
> having BlockBasedTableIterator return a sentinel key for the beginning/end of a file when it's extended by range tombstones so it'll be closed at the same time the tombstones become irrelevant

I was thinking this would be done by `LevelIter` since it has a handle on the `FileMetadata`, though I suppose we can plumb that into `BlockBasedTableIterator`.